File: lab/explorer/cpp/immutability/immutability_advanced.cpp
#include <iostream>

struct Point {
    int x;
    int y;
};

int main() {
    Point p{1, 2};

    // 1) pointer to const data (you can't change *p1, but you can move p1)
    const Point* p1 = &p;
    // p1->x = 10;        // ❌ error: data is const
    p1 = nullptr;          // ✅ pointer itself is movable

    // 2) const pointer to mutable data (you can change *p2, but not p2)
    Point* const p2 = &p;
    p2->x = 20;            // ✅ ok: data is mutable
    // p2 = nullptr;       // ❌ error: pointer is const

    // 3) const pointer to const data (neither)
    const Point* const p3 = &p;
    // p3->x = 30;         // ❌ cannot modify data
    // p3 = nullptr;       // ❌ cannot reassign pointer

    std::cout << "p = (" << p.x << ", " << p.y << ")\n";
    return 0;
}

File: lab/explorer/rust/immutability/immutability_advanced.rs
struct Point {
    x: i32,
    y: i32,
}

fn main() {
    // immutable binding → whole value is frozen
    let _p = Point { x: 1, y: 2 };
    // _p.x = 10;              // ❌ cannot assign to field of immutable binding

    // mutable binding → whole value can change
    let mut q = Point { x: 3, y: 4 };
    q.x = 30;                  // ✅

    // immutable borrow
    let _r: &Point = &q;
    // _r.x = 99;               // ❌ cannot modify through &T

    // mutable borrow
    let r_mut: &mut Point = &mut q;
    r_mut.x = 50;              // ✅

    // if you want C++-style "interior mutability", you must opt in:
    use std::cell::Cell;
    struct Wrap {
        inner: Cell<i32>,
    }
    let w = Wrap { inner: Cell::new(10) };
    // w is not mut, but inner can change:
    w.inner.set(42);           // ✅ interior mutability

    println!("q = ({}, {})", q.x, q.y);
    println!("w.inner = {}", w.inner.get());
}

